### PR TITLE
Define feature gates centrally and document relevance for components

### LINKED ISSUE #5486: Define feature gates centrally instead of once per component
**How to categorize this issue?**

/area dev-productivity quality
/kind cleanup

**What would you like to be added**:

We define feature gates multiple times, once per component which uses it.

E.g. the `UseDNSRecords` feature gate is defined three times:
https://github.com/gardener/gardener/blob/744e0ff5db4f7f58d70a2f43546d379303f8a2cf/pkg/apiserver/features/features.go#L28
https://github.com/gardener/gardener/blob/744e0ff5db4f7f58d70a2f43546d379303f8a2cf/pkg/controllermanager/features/features.go#L29
https://github.com/gardener/gardener/blob/744e0ff5db4f7f58d70a2f43546d379303f8a2cf/pkg/gardenlet/features/features.go#L35

Also, there is one `featuregate.NewFeatureGate()` call per component, i.e. one feature gate map per component.

Instead, feature gates should be defined centrally once and added to `utilfeature.DefaultMutableFeatureGate` like in k/k:

https://github.com/kubernetes/kubernetes/blob/b181e174f25df752eb176c2393b95569646230d3/pkg/features/kube_features.go#L829
https://github.com/kubernetes/kubernetes/blob/b181e174f25df752eb176c2393b95569646230d3/pkg/features/kube_features.go#L822-L824

**Why is this needed**:

Redefining features gates is a source of programmer errors: one can easily forget to adapt the multiple definitions. Also shared code might be looking at features gates without someone noticing, that the feature gate is not registered with the component's map of feature gates.

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity documentation
/kind cleanup
/merge squash

**What this PR does / why we need it**:
As described in https://github.com/gardener/gardener/issues/5486#issuecomment-1060379518.

**Which issue(s) this PR fixes**:
Fixes #5486

**Release note**:

```doc operator
The [feature gate documentation](https://github.com/gardener/gardener/tree/master/docs/deployment/feature_gates.md) does now contain information about which of the feature gates are relevant for which Gardener components.
```

### CODE REVIEW

File: docs/deployment/feature_gates.md [hunk 1]
```diff
@@ -100,29 +100,25 @@ A *General Availability* (GA) feature is also referred to as a *stable* feature.
...
+| ForceRestore                               | `gardenlet`                                                      | Enables forcing the shoot's restoration to the destination seed during control plane migration if the preparation for migration in the source seed is not finished after a certain grace period and is considered unlikely to succeed (falling back to the [control plane migration "bad case" scenario](../proposals/17-shoot-control-plane-migration-bad-case.md)). If you enable this feature gate, make sure to also enable `UseDNSRecords` and `CopyEtcdBackupsDuringControlPlaneMigration`. |
+| DisableDNSProviderManagement               | `gardenlet`                                                      | Disables management of `dns.gardener.cloud/v1alpha1.DNSProvider` resources. In this case, the `shoot-dns-service` extension will take this over if it is installed. This feature is only effective if the feature `UseDNSRecords` is `true`. |
+| ShootCARotation                            | `gardener-apiserver`                                             | Enables the feature to trigger automated CA rotation for shoot clusters. |
```
Comment (shafeeqes) [inline docs/deployment/feature_gates.md]:
```suggestion
| ShootCARotation                            | `gardener-apiserver`, `gardenlet`                                             | Enables the feature to trigger automated CA rotation for shoot clusters. |
```

**Comment by shafeeqes** [comment]
Nice!
/lgtm

**Review by ialidzhikov — APPROVED** [review]
Thank you!

/lgtm

**Review by timebertt — COMMENTED** [review]
Thanks!
Deduplicating the specs was one motivation behind https://github.com/gardener/gardener/issues/5486.

However, I think

> Also shared code might be looking at features gates without someone noticing, that the feature gate is not registered with the component's map of feature gates.

is still relevant. How do we make sure to prevent such programmer errors, which are easy to miss?

Generally, my feeling is that feature gates should simply be configured in all components consistently (i.e. enabled everywhere or not at all) and no one should think or care about which of them has handling for it or not.
WDYT?

**Comment by rfranzke** [comment]
> Generally, my feeling is that feature gates should simply be configured in all components consistently (i.e. enabled everywhere or not at all) and no one should think or care about which of them has handling for it or not.
WDYT?

@timebertt We discussed your concerns beginning of this week but agreed on this solution (ref https://github.com/gardener/gardener/issues/5486#issuecomment-1060379518). 

> How do we make sure to prevent such programmer errors, which are easy to miss?
We still don't, but I also don't think that this is a huge problem (not even a small one, actually). Did we ever hit such case? Such programmer errors will easily reveal during development/testing, or?

**Comment by timebertt** [comment]
> We still don't, but I also don't think that this is a huge problem (not even a small one, actually). Did we ever hit such case? 

True, we have not hit such a case yet – I think.

> Such programmer errors will easily reveal during development/testing, or?

Hmm, at least only during manual testing. Unit tests typically mutate the feature gates map. If you forget to adapt the component's features gate map, this will only be revealed during manual tests.

But ok, I'm fine to continue like this, if you have already agreed on this approach 👍

**Pull request merged**